include/nbl/asset/utils/CSPIRVIntrospector.h

```diff
@@ -13,172 +17,657 @@
+						inline std::enable_if_t<C,size_t> getRuntimeSize(const size_t lastMemberElementCount) const
+						{
+							if (isLastMemberRuntimeSized())
+							{
+								const auto& lastMember = type->memberTypes()[type->memberCount-1];
+								assert(!lastMember.count.front().isSpecConstantID);
+								return sizeWithoutLastMember+lastMemberElementCount*type->memberStrides()[type->memberCount-1];
+							}
+							return sizeWithoutLastMember;
+						}
+
+						//! Use `getRuntimeSize` for size of the struct with assumption of passed number of elements.
+						//! Need special handling if last member is rutime-sized array (e.g. buffer SSBO `buffer { float buf[]; }`)
+						size_t sizeWithoutLastMember;
+					};
+					struct SInputAttachment final
+					{
+						uint32_t index;
+					};
+
+					inline bool isArray() const {return !count.empty();}
+					inline bool isRunTimeSized() const {return isArray() ? count[0].value == 0 : false;}
+
+					//! Note: for SSBOs and UBOs it's the block name
+					span_t<char,Mutable> name = {};
+					span_t<SArrayInfo,Mutable> count = {};
+					uint8_t restrict_ : 1 = false;
+					uint8_t aliased : 1 = false;
+					
+					//
+					union
+					{
+						SCombinedImageSampler combinedImageSampler;
+						SStorageImage storageImage;
+						SUniformTexelBuffer uniformTexelBuffer;
+						SStorageTexelBuffer storageTexelBuffer;
+						SUniformBuffer uniformBuffer;
+						SStorageBuffer storageBuffer;
+						SInputAttachment inputAttachment;
+						// TODO: acceleration structure?
+					};
+				};
+
+				//! For the Factory Creation
+				struct SParams
 				{
-					auto found = std::find_if(descSet.begin(), descSet.end(), [](const SShaderResourceVariant& bnd) { return bnd.descCountIsSpecConstant; });
-					if (found != descSet.end())
-						return false;
+					std::string entryPoint;
+					core::smart_refctd_ptr<const ICPUShader> shader;
+
+					bool operator==(const SParams& rhs) const
+					{
+						if (entryPoint != rhs.entryPoint)
+							return false;
+						if (!rhs.shader)
+							return false;
+						if (shader->getStage() != rhs.shader->getStage())
+							return false;
+						if (shader->getContentType() != rhs.shader->getContentType())
+							return false;
+						if (shader->getContent()->getSize() != rhs.shader->getContent()->getSize())
+							return false;
+						return memcmp(shader->getContent()->getPointer(), rhs.shader->getContent()->getPointer(), shader->getContent()->getSize()) == 0;
+					}
+				};
+				inline const auto& getParams() const {return m_params;}
+
+				//! TODO: Add getters for all the other members!
+				inline const auto& getDescriptorSetInfo(const uint8_t set) const {return m_descriptorSetBindings[set];}
+				inline const auto& getInputs() const { return m_input; }
+				inline const core::vector<SFragmentOutputInterface>& getFragmentShaderOutputs() const
+				{
+					if (m_shaderStage != IShader::ESS_FRAGMENT)
+					{
+						// TODO: log error
+						return {};
```

Comment:
you don't want to fail/bail from a method/function that returns a reference, by returning a reference to a temporary

you either return a pointer, or if the function happens to return a `const&` like here, you return a reference to a `const static`

Reply:
returning a span would probably work better